### ISSUE TITLE
[Bug] Qwen3-TTS 0.6B Base fails to emit codec EOS and runs to max_new_tokens (2048 tokens / 163.84 s)

### ISSUE THREAD

**Issue by lijrjyan** [sgl-project/sglang-omni#1179]
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [x] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sglang-omni/discussions/new/choose Otherwise, it will be closed.
- [x] 5. Please use English, otherwise it will be closed.

### Describe the bug

Qwen3-TTS-12Hz-0.6B-Base intermittently fails to emit its codec EOS token. The model renders the target utterance completely and correctly, then enters a stationary near-silence state and generates room tone until `max_new_tokens` is reached — producing exactly 2048 tokens / 163.84 s of audio where ~50 tokens / ~4 s were expected.

Rate is **207 of 60,424 requests (0.34%)** on a full SeedTTS run (EN ≈0.7%, ZH ≈0.15%), flat across concurrency 1–32.

There are two separable problems here.

**1. The starvation itself is model-intrinsic.** Inside the state, EOS(2150) probability collapses to ~1e-14 (rank ~1500–1900, far outside `top_k=50`) and stays flat for 1,500 steps. EOS is never generated, so `repetition_penalty` never touches it — the collapse is the model's own distribution. Disabling the radix cache and wiping all JIT/compile caches does not change the rate, so this is not a cached-prefill or kernel correctness defect. Same signature as #599 and the #570 diagnosis.

**2. SGLang-Omni makes it silent.** The scheduler knows the request stopped on length, but `finish_reason` is not propagated to the Qwen3-TTS result adapter or the HTTP response, so `/v1/audio/speech` returns HTTP 200 with valid-looking audio and the benchmark scores all 207 failures as successes. This is the actionable engine-side bug.

Because no sampling parameter can recover a token at rank ~1500, mitigation has to be detection — a sustained low-energy stop condition or a duration-to-text-length bound — plus exposing `finish_reason` so callers can tell a length stop from an EOS stop.

Full analysis, statistics and discriminator results are in the first comment below.

### Reproduction

Model: `Qwen/Qwen3-TTS-12Hz-0.6B-Base` (rev `5d83992436eae1d760afd27aff78a71d676296fc`).

Server:

```bash
python -m sglang_omni.cli serve \
  --model-path <qwen3-tts-0.6b-base> \
  --model-name Qwen/Qwen3-TTS-12Hz-0.6B-Base \
  --config examples/configs/qwen3_tts_0_6b.yaml \
  --port 18006
```

Benchmark, over an 80-prompt set of 40 known-susceptible + 40 matched control SeedTTS EN prompts, repeated 4×:

```bash
python -m benchmarks.eval.benchmark_tts_seedtts \
  --use-existing-server --generate-only \
  --base-url http://127.0.0.1:18006 \
  --model Qwen/Qwen3-TTS-12Hz-0.6B-Base \
  --meta <meta.lst> --lang en \
  --max-concurrency 1 --warmup 1 \
  --seed 123 --max-new-tokens 2048 \
  --output-dir <out>
```

Sampling values are the model's own `generation_config.json` defaults: `temperature=0.9`, `top_p=1.0`, `top_k=50`, `repetition_penalty=1.05`.

**Expected:** all requests terminate on codec EOS(2150), `finish_reason=stop`, ~50 completion tokens.

**Actual:** 3 of 80 fail identically on all four passes, with `completion_tokens=2048`, `audio_duration_s=163.84`, no EOS generated, and HTTP 200:

| sample_id | text |
|---|---|
| `common_voice_en_17879420-common_voice_en_17879418` | Every cloud has a silver lining. |
| `common_voice_en_25572984-common_voice_en_25572982` | This was the same ratings received as the pre… |
| `common_voice_en_581024-common_voice_en_581027` | Epilepsy is a condition which sends the suffe… |

Reproduction notes:

- Run with `disable_radix_cache: true` and concurrency 1, and **do not clear the JIT caches** (`~/.triton`, `/tmp/torchinductor_root`, `~/.cache/flashinfer`, `~/.nv`). Under those conditions the failing set is bit-stable across passes.
- With the radix cache enabled, passes 2..N are stable but pass 1 differs (it populates the cache).
- Changing any of those — concurrency, cache state, compile/autotune state, driver version — keeps the **rate** and the **at-risk prompt pool** but moves *which* members fail. A single named prompt cannot be guaranteed to fail under an arbitrary configuration; see the comment below.

### Environment

| | |
|---|---|
| OS | Ubuntu 24.04.3 LTS |
| GPU | 1× NVIDIA GeForce RTX 5090, 32,607 MiB, SM120 |
| Driver / CUDA | 595.71.05 / 13.2 |
| PyTorch | 2.11.0+cu130 |
| SGLang-Omni | `043f9ae6bd10525eeec8c6d8bc114dbba3f7b731` (tree `de0737f11c37272e12998744fcd6d027cb7de000`) |
| Model | `Qwen/Qwen3-TTS-12Hz-0.6B-Base` rev `5d83992436eae1d760afd27aff78a71d676296fc` |
| Dataset | SeedTTS rev `27f4c1adee83b5b29b7c4b375f6b976324bda308` |

A second RTX 5090 (driver 580.95.05 / CUDA 13.0) was tested but only over a 102-request window, which is too small to be informative at this rate (expected ~1.4 events); that run is inconclusive, not a negative result.

### Related

- #599 — roadmap item, same signature
- #570 — prior diagnosis as a stochastic decode tail
- [QwenLM/Qwen3-TTS#118](https://github.com/QwenLM/Qwen3-TTS/issues/118) — rare missing-EOS in pretrained Base models; closed inactive without a root-cause fix

**Comment by lijrjyan** [comment]
## Root-cause analysis

This is the evidence behind the report above. It also directly answers the open TODO on the roadmap item in #599 ("evaluate alt `top_k` / temperature / `repetition_penalty` / retry-on-length") — see the last section.

All work is on a single frozen stack: SGLang-Omni `043f9ae6b` (tree `de0737f1`), model rev `5d839924`, RTX 5090 SM120, driver 595.71.05 / CUDA 13.2, seed 123.

### 1. Mechanism: an absorbing state, not a drift

EOS(2150) probability after penalty, per decode step, for one captured runaway:

| step | 128 | 256 | 512 | 1024 | 1536 | 2047 |
|---|---|---|---|---|---|---|
| P(EOS) | 5.5e-05 | 3.5e-11 | 1.5e-15 | 5.1e-14 | 4.3e-14 | 2.1e-14 |
| rank | 18 | 350 | 1935 | 1504 | 1565 | 1772 |

The probability collapses ten orders of magnitude by step 512, then **plateaus flat for the remaining 1,500 steps**. Drift or accumulating numerical error would keep moving; a flat plateau is a stationary distribution — an absorbing limit cycle.

Critically, **EOS is never generated, so it is never touched by `repetition_penalty=1.05`**. The collapse is entirely the model's own conditional distribution, not an artifact of the penalty.

The audio identifies the state. For the runaway rendering of `common_voice_en_652024-common_voice_en_652023` ("Add autry inman to my Hanging Out and Relaxing playlist."), RMS per 0.5 s:

```
0.016, 0.050, 0.052, 0.014, 0.068, 0.068, 0.031, 0.053, 0.008, 0.0004, 0.0008, 0.0016, …
```

The sentence is spoken **completely and cleanly in ~4.5 s**; the remaining **159 s is stationary near-silence** at RMS ≈0.0023 (≈ −53 dBFS). Corrupted KV would produce loud garbage. Clean speech followed by stable room tone is a model that finished talking and cannot say "done".

### 2. Susceptibility is a property of the prompt

From a full SeedTTS run — 60,424 accepted requests, 44 arms (EN/ZH × concurrency 1–32 × 3 repeats):

- **207/60,424 = 0.3426%** overall (EN ≈0.7%, ZH ≈0.15%).
- The rate is **flat across concurrency 1→32** and reproduces in every arm. A batching or cache-reuse defect would not be concurrency-insensitive.

Permutation test (20,000 resamples, statistic = pair-coincidences of runaway `sample_id`s across arms):

| | events | distinct ids | observed | null | z |
|---|---:|---:|---:|---|---:|
| EN | 158 | 75 | 149 | 10.9 ± 3.3 | **41.8** |
| ZH | 49 | 32 | 25 | 0.5 ± 0.7 | **33.2** |

Clustering by **reference speaker** is equally strong (EN z = 40.5): `common_voice_en_27099623` alone accounts for 10 of 158 EN events out of 666 speakers.

This is not a length artifact — ever-runaway and never-runaway prompts have near-identical length distributions (median normal completion 50 vs 51 tokens), and the rate is non-monotonic across length deciles.

### 3. Discriminator: it is not the prefix cache, and not compiled-kernel state

Enriched design: 40 EN prompts with ≥2 hits across the 22 EN arms ("susceptible") + 40 zero-hit, length-matched prompts ("control"). Fresh server, concurrency 1, seed 123, 4 repeats per arm.

| arm | susceptible | control |
|---|---|---|
| radix **ON** | 17/160 = **10.62%** [6.74, 16.36] | **0/160** |
| radix **OFF** | 12/160 = **7.50%** [4.34, 12.65] | **0/160** |
| radix **OFF**, JIT/compile caches wiped | 16/160 = **10.00%** [6.25, 15.63] | **0/160** |

Three-arm rate homogeneity **chi² = 1.030, p = 0.598** — indistinguishable. Pooled susceptible **45/480 = 9.38%** vs control **0/480**, Fisher **p = 1.9e-14**.

Ablation effectiveness was verified at the runtime level, not just in config:

| arm | prefill behavior in server log |
|---|---|
| radix ON | `#new-token: 1, #cached-token: 59` — prefix reuse active |
| radix OFF | `#new-token: 58, #cached-token: 0` — full prefill every request |

**Every request in the radix-OFF arms took the full-prefill route, and the failure still occurred at the same rate.** The JIT-wiped arm additionally discarded all Triton / FlashInfer / Inductor / NVIDIA compute caches (8,317 files) and forced full recompilation and fresh CUDA graph capture. Neither changed the rate, and control stayed at zero in all three arms.

### 4. Where the non-determinism actually comes from

`sglang/srt/layers/sampler.py` uses `multinomial_with_seed(logprobs, sampling_seed, positions)` — sampling is a deterministic function of (request seed, token position), independent of batch composition. So **the only run-to-run variation is the logits themselves**: floating-point non-associativity in the forward pass, whose reduction order depends on batch composition, prefill route, CUDA graph replay, and kernel autotune selection.

This predicts the observed determinism structure, which was confirmed:

| condition | repeats | observed |
|---|---|---|
| radix ON, c1 | r1 vs r2/r3/r4 | **disjoint** (r1 is the cold-cache pass) |
| radix ON, c1 | r2, r3, r4 | **identical** |
| radix OFF, c1 | r1–r4 | **all four identical** |
| radix OFF, JIT wiped | r1–r4 | **all four identical** |
| full run, EN c1 | r2, r3 | identical (10 prompts) |
| full run, ZH c1 | r1, r2, r3 | identical (3 prompts) |
| full run, EN c4/c8/c16/c32 | r1, r2, r3 | differ (batch composition varies) |

Two independent runaway requests produced EOS probability curves agreeing to the last decimal digit (`5.4580414143856615e-05`) and byte-identical 163.84 s WAVs.

Wiping the JIT caches changed the failing set **completely — zero overlap** — while the rate held and control stayed at zero:

| run | failing prompts |
|---|---|
| radix OFF, warm JIT | `17879418`, `25572982`, `581027` |
| radix OFF, JIT wiped | `20918611`, `25014863`, `27099625`, `96263` |

Across all three arms, 12 distinct prompts ever failed: **all 12 inside the 40-prompt susceptible pool, none in control**; pairwise membership overlap 2, 2, 0. The four new members are the highest-susceptibility ones (`27099625` is the global maximum at 6/22).

So: **the model fixes the at-risk pool and the rate; floating-point state decides which pool members trip on a given pass.**

### 5. What this means for the #599 TODO

The roadmap item proposes evaluating alternative `top_k` / temperature / `repetition_penalty` / retry-on-length. Based on the above:

- **`top_k` widening cannot work.** Once inside the absorbing state EOS sits at rank ~1500–1900. Recovering it would require `top_k` ≳ 2000 on a ~2150-entry codec vocabulary, i.e. effectively disabling top-k for every request.
- **`repetition_penalty` tuning cannot work.** EOS is never generated, so the penalty never applies to it. Changing the penalty only reshapes the non-EOS tail.
- **Temperature is a selector, not a fix.** It changes which trajectories enter the state and at what rate, but the state itself remains reachable. It would also change voice quality on the other 99.7% of requests.
- **Retry-on-length is viable** but currently impossible to implement at the API layer, because the length stop is not observable — see below.
- **Detection is the practical mitigation.** The state is stationary at ≈ −53 dBFS, so a sustained low-energy stop condition or a duration-to-text-length bound both terminate it cleanly and early.

Any evaluation of the above should use both a fixed-seed bad case and a prefix-cache-perturbed case, since in real serving the outlier is reached through runtime perturbation rather than a fixed seed.

### 6. The engine-side bug worth fixing independently

The scheduler has the correct termination state: `req.finished_reason` is set and copied into `data.finish_reason` before `_result_adapter(data)` is called in `sglang_omni/scheduling/omni_scheduler.py`. The Qwen3-TTS result adapter in `sglang_omni/models/qwen3_tts/request_builders.py` then constructs a new `StagePayload.data` containing audio codes and usage only — `finish_reason` is not carried. `Qwen3TTSState` has no finish field, and `SpeechResult` in `sglang_omni/client/types.py` has audio, format, sample rate and usage but no finish attribute.

Net effect: `/v1/audio/speech` returns HTTP 200 with a decodable WAV for a length-terminated generation, indistinguishable from a normal EOS stop. A deliberate `max_new_tokens=1` probe also returns HTTP 200 audio. This is why all 207 failures scored as successes in the benchmark, and why retry-on-length cannot be implemented by a caller today.

This is independent of the model defect and fixable on its own.

### 7. Reproduction boundary

Deterministic, and what is not:

- **Deterministic:** radix cache off, concurrency 1, fixed seed, JIT caches preserved → identical failing set on every pass (verified 4/4).
- **Not deterministic:** which prompts fail, across configurations. The three arms above shared only 2, 2 and 0 members.
- **Stable but noisy in small windows:** the rate (~0.7% EN / ~0.15% ZH) is a stable population property, but any 100-request window is Poisson noise around ~1 event. Small-window ablations at this rate are uninformative — a 102-request arm has `P(observe 0) ≈ 26%` even when the intervention does nothing.
- **Not bit-exact across hosts:** driver/CUDA changes kernel selection and reduction order, so two hosts agree on the rate and the at-risk pool but not on which members fail.

| uncertainty source | effect | controllable |
|---|---|---|
| model's absorbing state | sets at-risk pool and base rate | **no — model fix required** |
| bf16 reduction order (batch composition) | selects members, pass to pass | yes: concurrency 1 |
| prefix-cache state | selects members, cold vs warm | yes: `disable_radix_cache` |
| CUDA graph / compile / autotune | selects members, per server instance | yes: pin flags, preserve JIT caches |
| driver / CUDA version | selects members, per host | no (pin the host) |
| sampler RNG | none — seed+position deterministic | already deterministic |

Upstream context: [QwenLM/Qwen3-TTS#118](https://github.com/QwenLM/Qwen3-TTS/issues/118) reported rare missing-EOS in the pretrained Base models and was closed as inactive without a root-cause fix. A durable fix belongs in the model or its fine-tuning; the enriched deterministic reproduction above can validate one.

Traces, per-arm archives and the enriched prompt set can be shared if useful.

**Comment by lijrjyan** [comment]
> Looked into this locally: The report gives a clear expected-vs-actual bug, reproducible commands, precise symptoms, rates, environment, and an actionable finish_reason propagation gap. Would a focused PR for this be useful?

Thanks for looking into it! I don’t think a focused PR is necessary. The root cause appears model-side, while a robust framework-level mitigation would be complex, risky, and likely model-specific. I would keep this issue as documentation.
